espressostreamer/espresso_streamer.go

```diff
@@ -96,33 +99,62 @@ func NewEspressoStreamer(namespace uint64, hotshotUrls []string,
 		l1Reader:                      l1Reader,
 		espressoTEEVerifierABI:        espressoTEEVerifierAbi,
 		espressoTEEVerifierAddr:       espressoTEEVerifierAddress,
+		// Initally, we set the confirmed message pos to max uint64
+		// because we dont know the confirmed message pos yet
+		confirmedMessagePos: abi.MaxUint256.Uint64(),
 	}
 }
 
-/*
-Pop the first message from the queue
-It will return nil if the queue is empty
-*/
-func (s *EspressoStreamer) PopMessageWithMetadataAndPos() *MessageWithMetadataAndPos {
-	if len(s.messageWithMetadataAndPos) == 0 {
-		return nil
+func (s *EspressoStreamer) Refresh(ctx context.Context, fetchLastMessageAndEspressoBlock func() (uint64, uint64, error)) (bool, error) {
+	lastMessagePos, lastEspressoBlock, err := fetchLastMessageAndEspressoBlock()
+	if err != nil {
+		return false, err
+	}
+
+	log.Info("Refreshing espresso streamer", "lastMessagePos", lastMessagePos, "lastEspressoBlock", lastEspressoBlock, "confirmedMessagePos", s.confirmedMessagePos)
+	// If the last message pos only increase by 1, we dont need to refresh
+	// because this is expected behavior
+	if lastMessagePos == s.confirmedMessagePos+1 {
+		s.confirmedMessagePos = lastMessagePos
+		s.currentMessagePos = lastMessagePos + 1
+		return false, nil
 	}
-	messageWithMetadataAndPos := s.messageWithMetadataAndPos[0]
-	s.messageWithMetadataAndPos = s.messageWithMetadataAndPos[1:]
-	return messageWithMetadataAndPos
+
+	// If last message pos didnt increase by and if not equal to the confirmed message pos
+	// something went wrong
+	// we reset the values to the values returned from the source of truth.
+	if lastMessagePos != s.confirmedMessagePos {
+		s.confirmedMessagePos = lastMessagePos
+		s.currentMessagePos = lastMessagePos + 1
+		s.nextHotshotBlockNum = lastEspressoBlock
+		s.messageWithMetadataAndPos = []*MessageWithMetadataAndPos{}
+		return true, nil
+	}
+
+	return false, nil
 }
 
-/*
-Peek the first message from the queue
-It will return nil if the queue is empty
-This will be used to check if the the EspressoStreamer has recieved
-the next expected message in sequence
-*/
-func (s *EspressoStreamer) PeekMessageWithMetadataAndPos() *MessageWithMetadataAndPos {
-	if len(s.messageWithMetadataAndPos) == 0 {
-		return nil
+func (s *EspressoStreamer) Next() (*MessageWithMetadataAndPos, error) {
+
+	// Find if the current message pos is in the queue
+	// delete any messages before the current message pos
+	for i := 0; i < len(s.messageWithMetadataAndPos); i++ {
+		if s.messageWithMetadataAndPos[i].Pos < s.currentMessagePos {
+			s.messageWithMetadataAndPos = s.messageWithMetadataAndPos[i+1:]
```

Comment:
I think some bugs would occur here as you have changed the slice while `i` is increasing. Maybe iterate from the end of the slice?

Reply:
That is true, I will play around this code and see if this will cause issues

Reply:
https://github.com/EspressoSystems/nitro-espresso-integration/pull/479/commits/595ae1b54bcb51c1c029f072a43cda6f24602c53